include/SDL3/SDL_dialog.h

```diff
@@ -0,0 +1,175 @@
+ * `name` is a user-readable label for the filter (for example, "Office document").
+ *
+ * `pattern` is a semicolon-separated list of file extensions (for example,
+ * "doc;docx").
+ *
+ * \sa SDL_DialogFileCallback
+ * \sa SDL_ShowOpenFileDialog
+ * \sa SDL_ShowSaveFileDialog
+ * \sa SDL_ShowOpenFolderDialog
+ */
+typedef struct
+{
+    const char* name;
+    const char* pattern;
+} SDL_DialogFileFilter;
+
+/**
+ * Callback used by file dialog functions.
+ *
+ * The specific usage is described in each function.
+ *
+ * \sa SDL_DialogFileFilter
+ * \sa SDL_ShowOpenFileDialog
+ * \sa SDL_ShowSaveFileDialog
+ * \sa SDL_ShowOpenFolderDialog
+ */
+typedef void(*SDL_DialogFileCallback)(const char* const*, void*);
+
+/**
+ * Displays a dialog that lets the user select a file on their filesystem.
+ *
+ * This function should only be invoked from the main thread.
+ *
+ * This is an asynchronous function; it will return immediately, and the result
+ * will be passed to the callback.
+ *
+ * The callback will be invoked with a null-terminated list of files the user
+ * chose. The list will be empty if the user canceled the dialog, and it will
+ * be NULL if an error occured.
+ *
+ * Note that the callback may be called from a different thread than the one
+ * the function was invoked on.
+ *
+ * If an error occurs, the error can be obtained with SDL_GetError().
+ *
+ * Depending on the platform, the user may be allowed to input paths that don't
+ * yet exist.
+ *
+ * \param callback The function to be invoked when the user selects a file and accepts, or the user cancels the dialog, or an error occurs. The first argument is a null-terminated list of C strings, representing the paths chosen by the user. The list will be empty if the user canceled the dialog, and it will be NULL if an error occured. If an error occured, it can be fetched with SDL_GetError(). The second argument is the userdata pointer passed to the function.
+ * \param userdata An optional pointer to pass extra data to the callback when it will be invoked.
+ * \param window The window to bind the dialog to. May be NULL. Not all platforms support this option.
+ * \param filters A null-terminated list of SDL_DialogFileFilter's. May be NULL. Not all platforms support this option, and platforms that do support it may allow the user to ignore the filters.
+ * \param default_location The default location to start the dialog at. May be NULL. Not all platforms support this option.
+ * \param allow_many If non-zero, the user will be allowed to select multiple entries. Not all platforms support this option.
+ *
+ * \since This function is available since SDL 3.0.0.
+ *
+ * \sa SDL_DialogFileFilter
+ * \sa SDL_DialogFileCallback
+ * \sa SDL_ShowSaveFileDialog
+ * \sa SDL_ShowOpenFolderDialog
+ */
+extern DECLSPEC void SDLCALL SDL_ShowOpenFileDialog(SDL_DialogFileCallback callback, void* userdata, SDL_Window* window, const SDL_DialogFileFilter *filters, const char* default_location, int allow_many);
```

Comment:
It might be a good idea to allow these functions to return an error if creating the dialog fails. That way, it would be possible for a the application to display a generic fallback if the system version isn't available.

Reply:
Good idea, feel free to submit a PR for this.

Reply:
I do remember thinking about doing this at first, but I found that it would cause a few problems.

For many OSes, any code invoking the dialog needs to happens in a different thread, which makes it impossible to return any error code in case of a problem. Although it would be possible to do so on certain OSes, I preferred to keep it consistent across platforms to avoid requiring developers to have multiple different error handlers for different platforms, which is tedious but also opens the risk of forgetting to test for a certain platform.

I can change the return type anyways if it is still considered preferable, but I wanted to share my thoughts before changing the code.